design/bytecode.md

```diff
@@ -0,0 +1,106 @@
+# Byte code
+
+The Byte code of Seed is generated from the SeedIR code, and executed by the
+SeedInterpreter as a virtual machine. The byte code instructions are 32 bits in
+length. All instructions have an opcode in the first 7 bits, so there are at
+most 128 kinds of opcode. Other fields of the instructions are listed below:
+
+- A (8 bits): The operand A
+- B (8 bits): The operand B
+- C (8 bits): The operand C
+- K (1 bit): Indicates if the operand C is the index of the register (K = 0) or
+  constant (K = 1)
+- Bx (17 bits): The operand Bx, combined with B, C and K
+- sBx (17 bits): The operand sBx, A signed argument in the range of [-Max, Max)
+  (Max = 2 ^ 16). The actual value is the unsigned value (Bx) minus Max, so that
+  -Max is represented by 0, and Max - 1 is represented by 2 \* Max - 1.
+
+The layout of the instructions is illustrated as below:
+
+```text
+
+     3 3 2 2 2 2 2 2 2 2 2 2 1 1 1 1 1 1 1 1 1 1 0 0 0 0 0 0 0 0 0 0
+     1 0 9 8 7 6 5 4 3 2 1 0 9 8 7 6 5 4 3 2 1 0 9 8 7 6 5 4 3 2 1 0
+    |K|      C (8)    |      B (8)    |     A (8)     |   Op (7)    |
+    |             Bx (17)             |     A (8)     |   Op (7)    |
+    |            sBx (17)             |     A (8)     |   Op (7)    |
+
+```
+
+SeedInterpreter is a register based virtual machine. There are maximum 255
+registers for each function. The local variables of a function is pre-allocated
+in the registers by the compiler. So the user could not define more than 255
+local variables for a certain function.
+
+Each function has its own constant list, which is generated by the compiler. The
+elements in the constant list are usually numbers and strings without
+duplication.
+
+## Instruction Notation
+
+Following notations are used to explain the behavior of the instructions.
+
+- PC: Program Counter
+- R[A]: Register at index A
+- R[B]: Register at index B
+- R[C]: Register at index C
+- K[N]: Element N in the constant list
+- RK[C]: Register at index C (when K = 0) or constant at index C (K = 1)
```

Comment:
It is confusing to have both the constant flag K and the RK[C] notation in the same context.

As I see in Lua's VM design, notation K is only used for "element n in the constant list". Thus, RK[C] is easy to understand in Lua's context.

In this design, you use K as a flag bit to indicate whether the operand C is constant or not. Given this context, K[Bx] looks still okay but "RK[C] ... when K = 0" looks a bit weird since K appears twice in the same line, with different meanings.

How about using another notation for the constant flag? E.g., KF, for "constant flag"?

Reply:
The reason of the confusion might be the difference between Lua 5.4 change, which is not documented. Current design follows the concept of Lua 5.4 design.

The instruction layout of Lua 5.0-5.3 is:
| op(6) | A(8) | B(9) | C(9) |

So both B and C support RK() operation to select register or constant based on the 9th bit. The binary operation should be:
ADD A B C         // R[A] := RK[B] + RK[C]

Lua 5.4 changes the layout to:
| op(7) | A(8) |K| B(8) | C(8) |

I think the intention is to double supported opcode numbers to support both ADDK and ADD, so that one condition check for RK() operation can be avoid. The binary operation should be:
ADDK A B C      // R[A] := R[B] + K[C]              K[B] + K[C] should be optimized into compile time calculation
ADD A B C        // R[A] := R[B] + R[C]

There are some instruction in Lua 5.4 uses RK() operation for array element and class self access. I'm not sure if RK() operation can be avoid now.

Maybe we can have several options:
1. Follow Lua 3.5's layout, and ignore Lua 5.4's optimization. It might not be a significant improvement.
2. Follow Lua 5.4's layout, but combine C and K to a 9 bits C for RK(C) support. ADDI is a very small optimization, may not needed.